src/UseCases/UpdateDonor/UpdateDonorUseCase.php

```
@@ -0,0 +1,85 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\DonationContext\UseCases\UpdateDonor;
+
+use WMDE\Fundraising\DonationContext\Authorization\DonationAuthorizer;
+use WMDE\Fundraising\DonationContext\Domain\Model\Donor;
+use WMDE\Fundraising\DonationContext\Domain\Model\DonorAddress;
+use WMDE\Fundraising\DonationContext\Domain\Model\DonorName;
+use WMDE\Fundraising\DonationContext\Domain\Repositories\DonationRepository;
+
+/**
+ * @license GNU GPL v2+
+ */
+class UpdateDonorUseCase {
+
+	private $authorizationService;
+	private $donationRepository;
+
+	public function __construct( DonationAuthorizer $authorizationService, DonationRepository $donationRepository ) {
+		$this->authorizationService = $authorizationService;
+		$this->donationRepository = $donationRepository;
+	}
+
+	public function updateDonor( UpdateDonorRequest $updateDonorRequest ): UpdateDonorResponse {
+		if ( !$this->requestIsAllowed( $updateDonorRequest )) {
+			return new UpdateDonorResponse( 'donor_change_failure_access_denied' );
+		}
+		$donation = $this->donationRepository->getDonationById( $updateDonorRequest->getDonationId() );
+
+		if ($donation->isExported()) {
+			return new UpdateDonorResponse( 'donor_change_failure_exported' );
+		}
+
+		try {
+			$donor = new Donor(
+				$this->getDonorNameFromRequest( $updateDonorRequest ),
+				$this->getDonorAddressFromRequest( $updateDonorRequest ),
+				$updateDonorRequest->getEmailAddress()
+			);
+		}
+		catch ( \UnexpectedValueException $e ) {
+			return new UpdateDonorResponse( 'donor_change_failure_generic' );
+		}
+
+		$donation->setDonor( $donor );
+		$this->donationRepository->storeDonation( $donation );
```

Please handle the `StoreDonationException`.

Or not, maybe it's ok to let it bubble up? Let's discuss.